ephemeral SSH key support

cluster2600 opened the pull request:
ephemeral SSH key support

dimonomid commented:
Also, regarding ssh config parsing specifically: I wonder if it'd make more sense to avoid adding this extra preprocessing and parsing logic right in Nerdlog, and instead improve the `ssh_config` library itself? I mean Nerdlog already uses my own fork https://github.com/dimonomid/ssh_config (which is actually also a fork of the original one), so I'll be more than happy to accept a good patch to implement `Include` and `Match` into this fork.

That library also has some test suite, so will probably be easier to add tests for this `Include` and `Match` thing there, instead of adding them to Nerdlog tests.

cluster2600 commented:
Hello @dimonomid 

Thank you for the thorough review! You've raised some excellent points. Let me address each of your concerns:

## 1. Separation of Concerns

You're absolutely right that this PR is doing two distinct things:
- **SSH Config improvements**: Adding support for `Include` and `Match` directives
- **Ephemeral key support**: Adding OIDC-based authentication via ephemeral SSH keys

I agree these should be separated. I'll create two separate PRs:
- **PR #1**: SSH Config improvements (Include and Match support)
- **PR #2**: Ephemeral key support (addressing issue #31)

This will make the changes easier to review and test independently.

## 2. README and CHANGELOG Replacement

You're correct - I accidentally replaced the entire README and CHANGELOG instead of updating them properly. Looking at the git history:
- The original README was a comprehensive project documentation (~100 lines)
- I replaced it with just the feature descriptions (~15 lines)
- Same issue with CHANGELOG - I removed all historical entries

This was indeed meant to be PR description content, not a replacement of the documentation. I'll restore the original files and add the new features as proper sections.

## 3. Testing Ephemeral Keys Locally

Great question about testing! Here's how to set up a minimal test environment:

### Local Testing Setup

1. **Mock Provider Testing**:
   - The codebase includes `ephemeral_key_provider_mock.go` which can be used for testing
   - You can enable it with: `--ephemeral-key-provider=mock`
   - This simulates the ephemeral key flow without requiring actual infrastructure

2. **Using opkssh locally**:
   ```bash
   # Install opkssh
   go install github.com/openpubkey/opkssh@latest
   
   # Set up a test OIDC provider (e.g., using Keycloak or a mock server)
   # Configure opkssh to use it
   
   # Run nerdlog with opkssh provider
   nerdlog --ephemeral-key-provider=opkssh --host=testserver
   ```

3. **Integration Tests**:
   - See `core/shell_transport_ssh_test.go` for existing test patterns
   - The mock provider allows testing the full flow without external dependencies

### GitHub Actions Testing

For CI/CD testing, I propose:
1. Use the mock provider for unit tests
2. Set up a containerized test environment with:
   - A mock OIDC provider
   - SSH server configured to accept ephemeral keys
   - Integration test suite

Example GitHub Actions workflow addition:
```yaml
ephemeral-key-tests:
  runs-on: ubuntu-latest
  steps:
    - uses: actions/checkout@v3
    - name: Set up test environment
      run: |
        docker-compose -f test/docker-compose.yml up -d
    - name: Run ephemeral key tests
      run: |
        go test -v ./core -run TestEphemeralKey
```

dimonomid commented:
Thanks for this detailed info @cluster2600 , very helpful! I definitely intend to play with opkssh and your changes on the weekend.

Regarding separation of PRs: you mentioned you were going to create a separate PR for the SSH changes, I just want to double check: would you mind making this change to the [ssh_config](https://github.com/dimonomid/ssh_config) library itself? So in Nerdlog we'll only need to update the version of that dependency, and maybe also make some minor changes if the API needs to be used differently (which it might, since `Include` requires reading multiple files as opposed to just consuming a single `Reader`)

cluster2600 commented:
## Thank You for the Excellent Suggestion!

You're absolutely right that contributing these improvements to the ssh_config library itself would be the better approach. This would benefit the entire Go community and result in cleaner architecture for Nerdlog.

## My Plan Moving Forward

I'll work on:

1. **Contributing the Include and Match directive support to the upstream ssh_config library**
   - Fork the upstream repository (likely `github.com/kevinburke/ssh_config`)
   - Port the Include directive preprocessing logic to work within the library
   - Port the Match directive filtering and evaluation
   - Add comprehensive tests and documentation
   - Submit a well-documented PR

2. **Once the upstream PR is merged, updating Nerdlog to use the new version**
   - Update the dependency to the new version
   - Remove the preprocessor code from Nerdlog
   - Adapt to any API changes if needed
   - Update PR #37 to simply bump the dependency version

3. **Keep the ephemeral key support PR (#38) as is**
   - This feature is specific to Nerdlog's functionality
   - It's ready for testing with the mock provider: `--ephemeral-key-provider=mock`

## Implementation Considerations

The current ssh_config library API works with a single `io.Reader`. With Include support, the API might need to:
- Accept a file path instead of just a Reader
- Or provide a callback mechanism for reading included files
- Or expose a preprocessing step that resolves includes

I'm thinking something like:
```go
// Current API
cfg, err := ssh_config.Decode(reader)

// Possible new API with Include support
cfg, err := ssh_config.DecodeFile("/path/to/ssh/config")
// or
cfg, err := ssh_config.DecodeWithIncludes(reader, includeResolver)
```

## Timeline

- **Week 1**: Prepare and submit PR to ssh_config library
- **Week 2-3**: Work with maintainers on review feedback
- **Week 4**: Update Nerdlog PR once upstream changes are merged

This approach is much cleaner and follows open source best practices. Thank you for pointing me in the right direction!

## Testing the Ephemeral Key Feature

In the meantime, you can easily test the ephemeral key feature without any setup:

```bash
# No infrastructure required - just run:
nerdlog --ephemeral-key-provider=mock --lstreams localhost
```

The mock provider generates test keys and simulates the ephemeral key workflow, so you can verify the feature works as expected.

Looking forward to your feedback on the ephemeral key implementation, and I'll keep you updated on the upstream ssh_config contribution progress!
